streaming/java/streaming-api/src/main/java/org/ray/streaming/python/descriptor/DescriptorFunction.java

```diff
@@ -0,0 +1,99 @@
+package org.ray.streaming.python.descriptor;
+
+import com.google.common.base.Preconditions;
+import org.ray.streaming.api.function.Function;
+
+/**
+ * A DescriptorFunction is used to describe a user defined python function.
+ * Python worker can create a function object based on information in this
+ * DescriptorFunction.
+ *
+ * <p>If DescriptorFunction is constructed from serialized python function,
+ * python worker can directly deserialize to create python function.
+ * If DescriptorFunction is constructed from moduleName and className/functionName,
+ * python worker will use `importlib` to load python
+ * function. </p>
+ *
+ * <p>When python data stream api call is from python, `pythonFunction` will be not null.</p>
+ * <p>When python data stream api call is from java, `moduleName` and `className`/`functionName`
+ * will be not null.</p>
+ */
+public class DescriptorFunction implements Descriptor, Function {
+  public enum PythonFunctionInterface {
+    SOURCE_FUNCTION("ray.streaming.function.SourceFunction"),
+    MAP_FUNCTION("ray.streaming.function.MapFunction"),
+    FLAT_MAP_FUNCTION("ray.streaming.function.FlatMapFunction"),
+    FILTER_FUNCTION("ray.streaming.function.FilterFunction"),
+    KEY_FUNCTION("ray.streaming.function.KeyFunction"),
+    REDUCE_FUNCTION("ray.streaming.function.ReduceFunction"),
+    SINK_FUNCTION("ray.streaming.function.SinkFunction");
+
+    private String pythonFunctionInterface;
+
+    PythonFunctionInterface(String pythonFunctionInterface) {
+      this.pythonFunctionInterface = pythonFunctionInterface;
+    }
+  }
+
+  private byte[] pythonFunction;
+  private String moduleName;
+  private String className;
+  private String functionName;
+  /**
+   * FunctionInterface can be used to validate python function,
+   * and look up operator class from FunctionInterface.
+   */
+  private String pythonFunctionInterface;
+
+  private DescriptorFunction(byte[] pythonFunction,
```

Comment:
How about use Either<L,R> style to refactor xlang pattern?

Reply:
It's mainly used in scala, not common style for java. And this constructor is private. User will use factory method.